### ISSUE TITLE
Ability to jump "through" one model to another on M:N relationships

### ISSUE THREAD

**Issue by cannikin** [prisma/prisma#2186]
## Problem

Once again this idea is inspired by Ruby on Rails. ;) ActiveRecord (the ORM library that ships with Rails) has the concept of a "has many through" relationship. That is, one record has many associated records **through** another record inbetween.

For example, let's say you have a `Car` which gets regular `Service`. The record of that car getting a service is handled by the `Maintenance` model, which records which car, which service, the date it was performed, and maybe the employee who performed the job. We could model this in `schema.prisma` like so (thanks to @pantharshit00 on Slack for working up the schema):

```prisma
model Car {
  id          Int        @default(autoincrement()) @id
  make        String
  model       String
  year        Int
  maintenance Maintenance[]
}

model Service {
  id          Int        @default(autoincrement()) @id
  name        String
  maintenance Maintenance[]
}

model Employee {
  id              Int        @default(autoincrement()) @id
  name            String
  maintenanceDone Maintenance[]
}

model Maintenance {
  id          Int      @default(autoincrement()) @id
  car         Car      @relation(fields: [carId], references: [id])
  service     Service  @relation(fields: [serviceId], references: [id])
  employee    Employee @relation(fields: [employeeId], references: [id])
  carId       Int
  serviceId   Int
  employeeId  Int
  performedAt DateTime @default(now())
}
```

Getting the `Maintenance` records on a `Car` is pretty easy:

```js
prisma.car.findOne({ where: { id: 123 } }).maintenance()
```

But what if I wanted the `Service` records associated with the car? Right now it would be something like:

```js
prisma.car.findMany({
    include: {
      maintenance: {
        select: {
          service: true,
        },
      },
    },
  })
```

But what if you could simplify that syntax with the idea of the "has many through" relationship relationship: A `Car` has many `Service` records through `Maintenance` records.

It would be amazing if you could get to those records using the same, simple syntax:

```js
prisma.car.findOne({ where: { id: 123 } }).service()
```

## Solution

Perhaps an extra parameter can be passed to `@relation` somehow denoting that the records are reached through another? You give it the name of another field you've already defined on the model, like `maintenance`:

```
model Car {
  id          Int        @default(autoincrement()) @id
  make        String
  model String
  year Int
  maintenance Maintenance[]
  service Service[] @relation(through: maintenance)
}
```

Since we already have the definition of how a `Service` relates to a `Maintenance` in the `Maintenance` model there's no need to repeat the other relation arguments.

## Additional context

Here's some Rails documentation on has-many-through: https://guides.rubyonrails.org/association_basics.html#the-has-many-through-association

What do you think? Is it crazy enough to work?

**Comment by arefaslani** [comment]
@cannikin Checkout this: https://www.prisma.io/docs/concepts/components/prisma-schema/relations#many-to-many-relations

**Comment by cannikin** [comment]
@arefaslani Thanks for the link! That's the standard many-to-many syntax, which is fine. For example, a `BlogPost` has many `Comments`, which works just fine: `const comments = db.blogPost.findOne({ where: { id: 1234 } }).comments()`

This introduces third (or more) levels of relationships, all accessible from the parent model, where maybe a `Comment` has one `Author`. What if I want to get all the comment authors via the `BlogPost`? That's where has-many-through comes into play. Prisma would have to add some kind of syntax to let me just do something like: `const authors = db.blogPost.findOne({ where: { id: 1234 } }).commentAuthors()`

**Comment by arefaslani** [comment]
@cannikin Yeah I encountered the same problem you described... I think it's a must have feature.

**Comment by umlx5h** [comment]
Simplified version in accordance with rails has_many through document.

schema
```
model Physician {
  id           Int           @id @default(autoincrement())
  name         String
  appointments Appointment[]
}

model Patient {
  id           Int           @id @default(autoincrement())
  name         String
  appointments Appointment[]
}

model Appointment {
  id           Int       @id @default(autoincrement())
  physician    Physician @relation(fields: [physician_id], references: [id])
  physician_id Int
  patient      Patient   @relation(fields: [patient_id], references: [id])
  patient_id   Int

  appointment_date DateTime @default(now())
}
```

insert records to all tables
```
  const patient_1 = await prisma.appointment.create({
    data: {
      patient: {
        create: { name: "patient_1" },
      },
      physician: {
        create: { name: "physician_1" },
      },
    },
  });
  await prisma.appointment.create({
    data: {
      patient: {
        create: { name: "patient_2" },
      },
      physician: {
        connect: { id: patient_1.id },
      },
    },
  });

```

To get patients served by a specific physician (=physician_1), 
```
  const physician = await prisma.physician.findFirst({
    where: { name: "physician_1" },
    include: { appointments: true },
  });

  const patients = await prisma.patient.findMany({
    where: {
      id: {
        in: physician.appointments.map((appointment) => appointment.patient_id),
      },
    },
  });

  console.log(patients);     => [ { id: 1, name: 'patient_1' }, { id: 2, name: 'patient_2' } ]
```

I want to above as follows without removing intermediate model from schema. Rails can do.
```
const physician = await prisma.physician.findFirst({
    where: { name: "physician_1" },
    include: { ??? },
  });

console.log(physician.patients);
```

**Comment by kenchambers** [comment]
we want has many through as well!

**Comment by mwebb33** [comment]
Bumping, has_many: through in rails is really nice. I noticed https://github.com/prisma/prisma/issues/7351 is similar to this ticket, but the attached PR is all-or-nothing for disabling fks. I want a specific has_many: through rails like syntax in Prisma for join table querying. It's kind of a pain querying without it.

**Comment by olefirenko** [comment]
Do not know if this helps but the Laravel framework also has the similar concept of `has many through` relationships https://laravel.com/docs/8.x/eloquent-relationships#has-many-through. Would be great to have the same feature in Prisma.

**Comment by jferrettiboke** [comment]
Any update on this? I'm also interested on this feature. Any ways we can help further? https://github.com/prisma/prisma/issues/4910

**Comment by mattiamalonni** [comment]
Just to keep thread alive.. This would be an awesome feature to have.
Also, really like the proposal approach.

**Comment by njwest** [comment]
+1, this would be a great feature to have. I'm coming to Prisma from Elixir-land and so far a `through` association is the only thing I'm missing (I am very thankful for Prisma's succinctness and its SQL migrations).

The `@relation(through: intermediaryRecord)` way of declaring the relationship on the schema is the clearest version of this to my mind, e.g. `service Service[] @relation(through: maintenance)` from @cannikin 's post.

Like in Rubyland, Elixir's Ecto schema syntax also has `:through` declarations for both `has_many` and `has_one` relationships, which are very helpful when writing concise queries for such data. 

From the [docs](https://hexdocs.pm/ecto/Ecto.Schema.html#has_many/3-has_many-has_one-through):

```elixir
  use Ecto.Schema

  schema "posts" do
    # In the has_many :through example below, the `:comments`
    # in the list [:comments, :author] refers to the
    # `has_many :comments` in the Post own schema and the
    # `:author` refers to the `belongs_to :author` of the
    # Comment's schema (the module below).
    # (see the description below for more details)
    has_many :comments_authors, through: [:comments, :author]

    # Specify the association with custom source
    has_many :tags, {"posts_tags", Tag}
  end

  schema "comments" do
    belongs_to :author, Author
    belongs_to :post, Post
    has_one :post_permalink, through: [:post, :permalink]
  end
```

1. I have a `Post` from the schema above that has_many `Comments`, and those `Comments` belong_to `Authors`
2. A single `post` has_many `:comments_authors` through `comments`

The two ways to take advantage of this when querying are:

1. Via a query `preload:` field, e.g. `preload: [:comments_authors]`, which in this case would return my `Post`, its  `CommentsAuthors`, and their `Comments` as if doing a comprehensive join under the hood.
2. Via querying only a record's associated `CommentsAuthors`, e.g. `authors = Repo.all assoc(post, :comments_authors)` would return only a post's `CommentsAuthors`.

**Comment by kidqueb** [comment]
Just to throw another idea in there to keep this alive. It's a little bit more verbose than prior suggestions, but keeps the data schema very opaque.

```
model User {
  id               Int               @id @default(autoincrement())
  groupMemberships GroupMembership[]
  groups           Group[]           @through(field: 'groupMemberships', references: 'group')
}

model Group {
  id Int @id @default(autoincrement())
}

model GroupMembership {
  id      Int   @id @default(autoincrement())
  user    User  @relation(fields: [userId], references: [id])
  userId  Int
  group   Group @relation(fields: [groupId], references: [id])
  groupId Int
}
```

```
const user = await client.user.findUnique({
   where: { id: 1 },
   include: { groups: true }
})
```

**Comment by arimendelow** [comment]
i’d also love this!! has anyone found any workarounds?

**Comment by DavidHooper** [comment]
+1 super keen on this

**Comment by devbytyler** [comment]
Also finding myself wanting a "has-one-through" relationship!

**Comment by lucassimines** [comment]
+1!

**Comment by maximbroos** [comment]
Would love this feature as well!

**Comment by dakotalightning** [comment]
+1 Keeping this alive :)

**Comment by apudiu** [comment]
I think the above proposals/ideas aren't perfect. In one of my Laravel project i've a `media` table. which stores all the files. product (different types of) images, videos, users avatars, store banner & logo etc. When I give a file to that, the model (before saving) automatically (ofcourse i've configured it) parses all the necessary (to project needs) meta data. like mime type, extension, file size, where the actual file is stored (system can store various types of files in different servers). When querying, this also construct appropriate full url of the file too.

**Comment by rbarisic-lme** [comment]
Is this still open?

**Comment by jamols09** [comment]
bump! :)

**Comment by BayenReyes** [comment]
+1 I need this

**Comment by dev-talha-anwar** [comment]
+1
